File: scala/Practice2/DynamicProgramming/Bit/Schedule.scala
// https://atcoder.jp/contests/joi2014yo/tasks/joi2014yo_d

object Main {

  implicit class Pipeline[T](x: T) {
    def |>[S](f: T => S): S = f(x)
  }

  def main(args: Array[String]) = (
    (
      io.StdIn.readLine.toInt,
      ("_" + io.StdIn.readLine)
        .map(c =>
          c match {
            case 'J' => 1
            case 'O' => 2
            case 'I' => 4
            case _   => 0
          }
        )
        .toArray
    )
    |> { case (n, rs) => (
      Array.range(0, n + 1).map(_x => Array.fill[Int](8)(0))
      |> (memo => (
        (memo(0)(1) = 1)
        |> (_x =>
          for (
            i <- Stream.range(1, n + 1);
            j <- Stream.range(1, 8)
              if (j & rs(i)) == rs(i);
            j2 <- Stream.range(1, 8)
              if (0 to 2).exists(x => ((j >> x & 1) == 1) && ((j2 >> x & 1) == 1))
          ) (
            memo(i)(j) = memo(i)(j) + memo(i - 1)(j2) % 10007
          )
        )
        |> (_x => memo(n).sum % 10007)
      ))
    )}
    |> println
  )
}

// https://atcoder.jp/contests/joi2014yo/submissions/20781702

File: fs/Practice2/DynamicProgramming/Bit/Schedule.fs
// https://atcoder.jp/contests/joi2014yo/tasks/joi2014yo_d

(
  stdin.ReadLine () |> int,
  "_" + stdin.ReadLine ()
  |> Seq.map (
    fun c ->
      match c with
      | 'J' -> 1
      | 'O' -> 2
      | 'I' -> 4
      | _   -> 0
  )
  |> Seq.toArray
)
|> fun (n, rs) ->
  Array.init (n + 1) (fun _ -> Array.create 8 0)
  |> fun memo ->
    (memo.[0].[1] <- 1)
    |> fun () ->
      seq {
        for i in 1..n do
        for j in 1..7 do
        if (j &&& rs.[i] = rs.[i])
        then yield! seq {
          for j2 in 1..7 do
          if (memo.[i - 1].[j2] > 0)
            && [0..2] |> List.exists (fun x -> (j >>> x &&& 1 = 1) && (j2 >>> x &&& 1 = 1))
          then yield (i, j, memo.[i - 1].[j2])
        }
      }
      |> Seq.iter (
        fun (i, j, v) ->
          memo.[i].[j] <- memo.[i].[j] + v % 10007
      )
      |> fun _ -> memo.[n] |> Array.sum |> fun v -> v % 10007
|> printfn "%d"

// ----------

// i in 1..n

// j: IOJ
// 0: 000
// 1: 001
// 2: 010
// 3: 011
// 4: 100
// 5: 101
// 6: 110
// 7: 111

// ----------

// input:
// 2
// OI

// memo:
// [|
//   [|0; 1; 0; 0; 0; 0; 0; 0|]
//   [|0; 0; 0; 1; 0; 0; 0; 1|]
//   [|0; 0; 0; 0; 1; 2; 2; 2|]
// |]

// output:
// 7

// https://atcoder.jp/contests/joi2014yo/submissions/20780648
